### PR TITLE
add adapter for opensea downpayment buy

### PR DESCRIPTION
**Pull request by akashiceth**
Draft now, will append tests later

### CODE REVIEW

File: contracts/adapters/OpenseaDownpaymentBuyAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    require(assets[0] == address(WETH), "Only WETH borrowing allowed");
+
+    // Check order payment token
+    address _buyerpaymentToken = _params.addrs[6];
+    address _sellerpaymentToken = _params.addrs[13];
+    require(_buyerpaymentToken == address(0), "Buyer payment token should be ETH");
+    require(_sellerpaymentToken == address(0), "Seller payment token should be ETH");
+
+    uint256 _flashBorrowedAmount = WETH.balanceOf(address(this));
+    assert(amounts[0] == _flashBorrowedAmount); // Should nerver happened
```
Comment (thorseldon) [inline contracts/adapters/OpenseaDownpaymentBuyAdapter.sol:85]: using require not asset

File: contracts/adapters/OpenseaDownpaymentBuyAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import {ILendPoolAddressesProvider} from "../interfaces/ILendPoolAddressesProvider.sol";
+import {IAaveLendPoolAddressesProvider} from "./interfaces/IAaveLendPoolAddressesProvider.sol";
+import {IWETH} from "../interfaces/IWETH.sol";
+import {IERC721Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721Upgradeable.sol";
+
+contract OpenseaDownpaymentBuyAdapter is
+  IAaveFlashLoanReceiver,
+  OwnableUpgradeable,
+  PausableUpgradeable,
+  ReentrancyGuardUpgradeable
```
Comment (thorseldon) [inline contracts/adapters/OpenseaDownpaymentBuyAdapter.sol:19]: need inherit ERC721Holder, for safeTransferFrom to addres(this)

File: contracts/adapters/OpenseaDownpaymentBuyAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    uint256 _buyerBalance = WETH.balanceOf(_buyer);
+    require(_buyerBalance + _flashBorrowedAmount >= _sellPrice, "Insufficient payment");
+
+    // Transfer buyer WETH for payment
+    uint256 _buyerPayment = _sellPrice - _flashBorrowedAmount;
+    WETH.transferFrom(_buyer, address(this), _buyerPayment);
+    WETH.withdraw(_sellPrice);
+
+    // Do opensea exchange
+    openseaExchange.atomicMatch_{value: _sellPrice}(
```
Comment (thorseldon) [inline contracts/adapters/OpenseaDownpaymentBuyAdapter.sol:104]: Is OpenSea use same api to deal with CryptoPunks?

Comment (akashiceth) [reply contracts/adapters/OpenseaDownpaymentBuyAdapter.sol:104]: No, must be NFT with ERC721 standard, WPunks is OK

File: contracts/adapters/OpenseaDownpaymentBuyAdapter.sol (hunk 1 above)
Comment (akashiceth) [reply contracts/adapters/OpenseaDownpaymentBuyAdapter.sol]: done

File: contracts/adapters/OpenseaDownpaymentBuyAdapter.sol (hunk 2 above)
Comment (akashiceth) [reply contracts/adapters/OpenseaDownpaymentBuyAdapter.sol]: done

**Pull request closed without merging**